WordPress/gutenberg
# Add a layout config to the group and theme.json and make alignments declarative

The way alignments work in Gutenberg right now is very optimized for traditional themes that provide their own alignment styles and it's also built to be backward compatible with themes that can or can't support wide alignments. This approach while it served us well in phase1, it can't scale properly to FSE because it's not really WYSWYG since the themes has to provide all the custom styles for all the alignments to work properly and it's also not possible to define different kind of alignments/widths in different contexts/templates/sidebars....

The current PR is a proof of concept of how we could make alignments (and layout in general) more declarative. It's based on the following principles:

 - An empty theme with no custom stylesheets should work properly and have the backend resemble the frontend without any custom CSS. (This PR will focus on alignments here but this principle is also broken for other reasons now)
 - Alignments and layout are context specific, which means there's a global layout but each container could potentially define its own internal layout for its children.
 - Based on the above, it becomes clear that the container blocks should have a way to define/override layouts.
 - Embrace the defaults from the browser: by default inserting "divs" just show take the whole width available, blocks should behave the same way unless the parent layout says otherwise.
 - Supporting wide/full widths and floats (left/right/center) at the same time consistently has proven to be a difficult task. The only way to achieve it properly is to actually alter the markup of blocks and add an extra wrapper. The idea in this PR is that block markup shouldn't be altered because of a properly which means this PR takes the opposite approach, instead of trying to hack these features together in the same container, it proposes that a container can authorize its children to have wide/full alignments or left/right ones but not both at the same time.

**Notes**

 - In terms of UI, the PR essentially adds controls to the sidebar of the group block, don't mind the ugly UI though, it's there only to show the capabilities, but it can be rethought in various ways including: having presets for layouts (or just a single default layout like the PR is doing right now), hide/show the custom properties based on theme configs...
 - The PR removes a bunch of CSS. Of course we can't just remove that CSS for backward compatibility reasons, so to move forward, we'd have to find a way to only limit the CSS to the existing containers and not to the one relying on the declarative approach.

**Testing the PR**

- Current themes (both block based and traditional) provide styles for different alignments... which means there will be conflicts, for this reason, I'm attaching an "empty" theme which almost nothing in its stylesheet for you to play with. 
[emptytheme.zip](https://github.com/WordPress/gutenberg/files/6041956/emptytheme.zip)

- Install the theme above
- Use the site editor (I didn't update the post editor to remove the opinionated styles for alignments/widths yet)
- Try modifying the templates by adding different group blocks, changing their widths, and aligning their inner blocks and see how things behave, if it feels natural and simple.


**Todo list**

- [x] Move the style generation to the server
- [x] Potentially remove the extra div in the group block (this div doesn't seem useful to me at all in this approach)
- [x] Check if we can build this as a support flag
- [x] Provide a way to inherit the layout of core/post-content block in the post editor
- [x] Think about the backward compatibility story, how can we introduce this behavior only to block based themes and ensure other themes keeping using the current behavior of the group block.
- [x] Polish theme.json integration
- [x] Design work to present this in the best way possible for container blocks.
- [ ] Follow-up: Update existing block-based themes as this PR might break some of them.
